book-2-platform-api/chapters/API_MODELS_CONTROLLERS.md

```diff
@@ -96,70 +158,121 @@ namespace CoffeeShop.Controllers
+                            BeanType = reader.GetString(reader.GetOrdinal("BeanType"))
+                        };
+                    }
+                    reader.Close();
+
+                    return Ok(coffee);
+                }
             }
         }
 
         [HttpPost]
         public async Task<IActionResult> Post([FromBody] Coffee coffee)
         {
-            string sql = $@"INSERT INTO Coffee
-            (Title, BeanType)
-            VALUES
-            ('{coffee.Title}', '{coffee.BeanType}');
-            select MAX(Id) from Coffee";
-
-            using (IDbConnection conn = Connection)
+            using (SqlConnection conn = Connection)
             {
-                var newId = (await conn.QueryAsync<int>(sql)).Single();
-                coffee.Id = newId;
-                return CreatedAtRoute("GetCoffee", new { id = newId }, coffee);
+                conn.Open();
+                using (SqlCommand cmd = conn.CreateCommand())
+                {
+                    cmd.CommandText = @"INSERT INTO Coffee (Title, BeanType)
+                                        OUTPUT INSERTED.Id
+                                        VALUES (@title, @beanType);
+                                        select MAX(Id) from Coffee";
+                    cmd.Parameters.Add(new SqlParameter("@title", coffee.Title));
+                    cmd.Parameters.Add(new SqlParameter("@beanType", coffee.BeanType));
+                    cmd.Parameters.Add(new SqlParameter("@id", id));
+
+                    cmd.ExecuteNonQuery();
+                    Int32 newId = (Int32) cmd.ExecuteScalar();
+                    coffee.Id = newId;
+                }
             }
-
+            return CreatedAtRoute("GetCoffee", new { id = newId }, coffee);
         }
 
         [HttpPut("{id}")]
         public async Task<IActionResult> Put([FromRoute] int id, [FromBody] Coffee coffee)
         {
-            string sql = $@"
-            UPDATE Coffee
-            SET Title = '{coffee.Title}',
-                BeanType = '{coffee.BeanType}'
-            WHERE Id = {id}";
-
             try
             {
-                using (IDbConnection conn = Connection)
+                using (SqlConnection conn = Connection)
                 {
-                    int rowsAffected = await conn.ExecuteAsync(sql);
-                    if (rowsAffected > 0)
+                    conn.Open();
+                    using (SqlCommand cmd = conn.CreateCommand())
                     {
-                        return new StatusCodeResult(StatusCodes.Status204NoContent);
+                        cmd.CommandText = @"UPDATE Coffee
+                                            SET Title = @title,
+                                                BeanType = @beanType
+                                            WHERE Id = @id";
+                        cmd.Parameters.Add(new SqlParameter("@title", coffee.Title));
```

Comment:
consider using `AddWithValue` like 
``` cs 
command.Parameters.AddWithValue("@title", coffee.Title);
```

Reply:
In the current rewrite that Andy's done for ADO.NET, he hasn't shown them `AddWithValue` yet, so I'm going to leave this with just `Add`, but we can start an issue about integrating `AddWithValue` into orientation.

Reply:
k

Reply:
I didn't know about `AddWithValue()`. How have I missed that all these years????

Reply:
yeah, I only found it about a year ago myself.  kicking myself for never knowing about it when i actually was using ado.net... :(